app/web_conn.go

```diff
@@ -304,19 +414,117 @@ func (wc *WebConn) writePump() {
 	}
 }
 
-// writeMessage is a helper utility that wraps the write to the socket
+// writeMessageBuf is a helper utility that wraps the write to the socket
 // along with setting the write deadline.
-func (wc *WebConn) writeMessage(msgType int, data []byte) error {
+func (wc *WebConn) writeMessageBuf(msgType int, data []byte) error {
 	wc.WebSocket.SetWriteDeadline(time.Now().Add(writeWaitTime))
 	return wc.WebSocket.WriteMessage(msgType, data)
 }
 
+func (wc *WebConn) writeMessage(msg *model.WebSocketEvent) error {
+	// We don't use the encoder from the write pump because it's unwieldy to pass encoders
+	// around, and this is only called during initialization of the webConn.
+	var buf bytes.Buffer
+	enc := json.NewEncoder(&buf)
+	err := msg.Encode(enc)
+	if err != nil {
+		mlog.Warn("Error in encoding websocket message", mlog.Err(err))
+		return nil
+	}
+	wc.Sequence++
```

Comment:
How does this relate to the increment in `writePump()`? Are they both necessary?

https://github.com/mattermost/mattermost-server/blob/54e1590b2bd276a8b3fcf19784556bf33ac92b12/app/web_conn.go#L363

Reply:
Yes absolutely. `wc.writeMessage` is only called before the for loop in the write pump actually starts.

Think of it this way:
- Client sends the last seen sequence number, which gets set in the constructor.
- We check for the dead queue, and either drain the queue, or send hello message. So sequence number needs to be incremented there.
- Then we come to our usual goroutine, where everything else remains same.

Reply:
Makes sense. I've got confused with the renaming and thought it was called further down in the function. All good then :+1: